### ISSUE TITLE
make Linux Kernel Runtime Guard (LKRG) easily avaialble in Qubes

### ISSUE THREAD

**Issue by adrelanos** [QubesOS/qubes-issues#5461]
**The problem you're addressing (if any)**
Yet unknown, upcoming kernel vulnerabilities.

**Describe the solution you'd like**
Linux Kernel Runtime Guard (LKRG)

**Where is the value to a user, and who might that user be?**
LKRG improves the security of the kernel. Therefore makes VM escapes harder.

The main argument for LKRG being: it renders whole classes of kernel exploits ineffective, makes other exploits less reliable / more difficult to write. LKRG was developed by a security professional with review from other high profile security professionals. References for that, you can read more about LKRG here:
https://www.whonix.org/wiki/Linux_Kernel_Runtime_Guard_LKRG

**Describe alternatives you've considered**

There's no real alternative as far as I know, just hardening coming from different approaches.

* https://github.com/Whonix/apparmor-profile-everything
* https://github.com/Whonix/hardened-vm-kernel
* https://github.com/anthraxx/linux-hardened

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**

* https://github.com/QubesOS/qubes-issues/issues/5456
* https://github.com/QubesOS/qubes-issues/issues/2844
* https://github.com/QubesOS/qubes-issues/issues/5212

----

Upstream is supportive. Links to communications with upstream, LKRG Debian packaging progress and development discussion can be found here:
https://forums.whonix.org/t/linux-kernel-runtime-guard-lkrg-linux-kernel-runtime-integrity-checking-and-exploit-detection/8477

Debian packaging status: The /debian folder is ready. I've successfully created a deb package on
my local machine. Managed figuring out using DKMS. The source code for that can be found here:
https://github.com/Whonix/lkrg

Instructions how to build the deb package from source code will come soonish. (Standard Debian methods / commands would do but this isn't trivial to research.) Binary packages will come soonish too.

Qubes status:

* Works for me in Qubes Debian buster template with VM kernel. (https://github.com/QubesOS/qubes-issues/issues/5212)
* Does not compile yet for me Qubes Debian buster template with dom0 kernel. Probably due to https://github.com/QubesOS/qubes-issues/issues/5456 / https://github.com/QubesOS/qubes-issues/issues/2844. This might be fixed after upgrades.
* Fedora templates / dom0 untested.

Might be prudent to make this an optionally installable pacakge at first. Then encourage wider testing. If all goes well, installation by default could be considered.

**Comment by adrelanos** [comment]
Packaging completed. Binary deb packages exist. Above post updated.

* [HowTo: Linux Kernel Runtime Guard (LKRG) in Qubes OS Debian or Qubes-Whonix ™ VMs](https://www.whonix.org/wiki/Linux_Kernel_Runtime_Guard_LKRG/Qubes)
* [HowTo: Build the Linux Kernel Runtime Guard (LKRG) Debian Package from Source Code](https://www.whonix.org/wiki/Linux_Kernel_Runtime_Guard_LKRG/Build_Debian_Package_from_Source_Code)

**Comment by fepitre** [comment]
If it helps with this by doing the Fedora side of packaging and testing.

**Comment by adrelanos** [comment]
Yes, sure.

**Comment by fepitre** [comment]
If you don't mind, I'm just assigning myself for tracking the issue for this part of the work.

**Comment by adrelanos** [comment]
I am OK with that. For me,  anything that moves this forward. :) (But I am not a github admin.)

**Comment by fepitre** [comment]
@adrelanos I've loaded the module into one of my Fedora based qube. Is there any particular tests I can execute to validate it works as expected? Maybe some malicious attacks or such?

**Comment by adrelanos** [comment]
I am not an expert for LKRG. Just a packager. I'd suggest to look that up upstream.

**Comment by fepitre** [comment]
@marmarek I was thinking to do like `dummy-psu` or `dummy-backlight`, having a separate component that would be built for templates (really straightforward) but including it as a submodule for `linux-kernel` without loading it by default. What do you think?

**Comment by DemiMarie** [comment]
This would be an especially good fit for KVM-based Qubes OS.

**Comment by marmarek** [comment]
> @marmarek I was thinking to do like `dummy-psu` or `dummy-backlight`, having a separate component that would be built for templates (really straightforward) but including it as a submodule for `linux-kernel` without loading it by default. What do you think?

Yes, I think that is a good idea. It makes it easier accessible when someone wants to use it. O

**Comment by fepitre** [comment]
I've prepared a Fedora version here: https://github.com/fepitre/fedora-lkrg/commit/8ec191895de4e4521f02b65c6abd93ca77b8c2d8. I'm waiting feedback from Fedora/RPMFusion people if they are interested to have it directly in their repositories.

**Comment by tommytran732** [comment]
@fepitre It has been awhile now... did they give you any feedback?

**Comment by fepitre** [comment]
> @fepitre It has been awhile now... did they give you any feedback?

Yes, it's just me lacking time to go into the whole integration process for a new package etc. I also would like to have more feedback from Fedora user people than just packaging it for one person.

**Comment by tommytran732** [comment]
Awesome!

**Comment by fepitre** [comment]
I've built a COPR repo with version 0.9.3 https://copr.fedorainfracloud.org/coprs/fepitre/lkrg/. You can install it by doing in a Fedora VM:
```
$ sudo dnf copr enable fepitre/lkrg
$ sudo dnf install lkrg-kmod
```

You may wait few secs and then
```
$ sudo systemctl start lkrg
```

Try systemctl restart if it fails the first time.

**Comment by andrewdavidwong** [comment]
> I've built a COPR repo with version 0.9.3 https://copr.fedorainfracloud.org/coprs/fepitre/lkrg/. You can install it by doing in a Fedora VM: [...]

@fepitre, does this mean that this issue is complete?

**Comment by tommytran732** [comment]
@fepitre Do you mind bumping the LKRG version on the COPR? It is a couple versions behind now :)

**Comment by fepitre** [comment]
> @fepitre Do you mind bumping the LKRG version on the COPR? It is a couple versions behind now :)

Sure. How things are going on Fedora with it?

**Comment by tommytran732** [comment]
> > @fepitre Do you mind bumping the LKRG version on the COPR? It is a couple versions behind now :)
> 
> Sure. How things are going on Fedora with it?

It works okay, though annoyingly `dnf copr` does not automatically get proxied through http://127.0.0.1:8082 in the template VM

**Comment by fepitre** [comment]
It's updated.
